plugin-server/src/kafka/consumer.ts

```diff
@@ -359,6 +361,45 @@ export class KafkaConsumer {
 
         const startConsuming = async () => {
             let lastConsumeTime = 0
+            const messageQueue: Message[] = []
+            let consumeTimer: NodeJS.Timeout | null = null
+            let isReadingInMainLoop = false
+
+            const continuousConsume = async (fromMainLoop = false) => {
+                if (!this.isStopping && (!fromMainLoop ? !isReadingInMainLoop : true)) {
+                    try {
+                        const batchSizeToFetch = fromMainLoop
+                            ? this.fetchBatchSize - messageQueue.length
+                            : Math.min(1, this.fetchBatchSize - messageQueue.length)
+                        if (batchSizeToFetch > 0) {
+                            const consumeStartTime = performance.now()
+                            if (lastConsumeTime > 0) {
+                                const intervalMs = consumeStartTime - lastConsumeTime
+                                histogramKafkaConsumeInterval.labels({ topic, groupId }).observe(intervalMs)
+                            }
+                            lastConsumeTime = consumeStartTime
+                            const newMessages = await retryIfRetriable(() =>
+                                promisifyCallback<Message[]>((cb) => this.rdKafkaConsumer.consume(batchSizeToFetch, cb))
+                            )
+                            messageQueue.push(...newMessages)
+                        }
+                    } catch (error) {
+                        logger.error('🔁', fromMainLoop ? 'main_loop_consume_error' : 'continuous_consume_error', {
+                            error: String(error),
+                        })
+                    }
+                } else if (this.isStopping) {
+                    if (consumeTimer) {
+                        clearInterval(consumeTimer)
```

Comment:
question: Looking at the exception handling code, I'm not sure how this cleanup is handled. We don't set `isStopping` on exceptions, only on clean shutdowns. I think the main loop will exit due to the exception thrown by the consumer after disconnection, but here we swallow the exceptions.